src/algorithms/model.ts

```diff
@@ -0,0 +1,376 @@
+  pop.infectionRate = (time: Date) =>
+    containment(time) * infectionRate(time.valueOf(), avgInfectionRate, params.peakMonth, params.seasonalForcing)
+
+  return pop
+}
+
+export function initializePopulation(
+  N: number,
+  numCases: number,
+  t0: number,
+  ages: Record<string, number>,
+): SimulationTimePoint {
+  // FIXME: Why it can be `undefined`? Can it also be `null`?
+  if (ages === undefined) {
+    const put = (x: number) => {
+      return { total: x }
+    }
+    return {
+      time: t0,
+      susceptible: put(N - numCases),
+      exposed: put(0),
+      infectious: put(numCases),
+      hospitalized: put(0),
+      critical: put(0),
+      discharged: put(0),
+      recovered: put(0),
+      intensive: put(0),
+      dead: put(0),
+      overflow: put(0),
+    }
+  }
+  const Z = Object.values(ages).reduce((a, b) => a + b)
+  const pop: SimulationTimePoint = {
+    time: t0,
+    susceptible: {},
+    exposed: {},
+    infectious: {},
+    hospitalized: {},
+    critical: {},
+    overflow: {},
+    discharged: {},
+    intensive: {},
+    recovered: {},
+    dead: {},
+  }
+  // TODO: Ensure the sum is equal to N!
+  Object.keys(ages).forEach((k, i) => {
+    const n = Math.round((ages[k] / Z) * N)
+    pop.susceptible[k] = n
+    pop.exposed[k] = 0
+    pop.infectious[k] = 0
+    pop.hospitalized[k] = 0
+    pop.critical[k] = 0
+    pop.overflow[k] = 0
+    pop.discharged[k] = 0
+    pop.recovered[k] = 0
+    pop.intensive[k] = 0
+    pop.dead[k] = 0
+    if (i === Math.round(Object.keys(ages).length / 2)) {
+      pop.susceptible[k] -= numCases
+      pop.infectious[k] = 0.3 * numCases
+      pop.exposed[k] = 0.7 * numCases
+    }
+  })
+
+  return pop
+}
+
+// Grabs all keys which index into a Record<string, number> object
+type NumberRecordKeys<T> = { [K in keyof T]: Record<string, number> extends T[K] ? K : never }[keyof T]
+
+// NOTE: Assumes all subfields corresponding to populations have the same set of keys
+export function evolve(pop: SimulationTimePoint, P: ModelParams, sample: (x: number) => number): SimulationTimePoint {
+  const sum = (dict: Record<string, number>): number => {
+    return Object.values(dict).reduce((a, b) => a + b, 0)
+  }
+  const fracInfected = sum(pop.infectious) / P.populationServed
+
+  const newTime = new Date(pop.time + P.timeDelta)
+  const newPop: SimulationTimePoint = {
+    time: newTime.valueOf(),
+    susceptible: {},
+    exposed: {},
+    infectious: {},
+    recovered: {},
+    hospitalized: {},
+    critical: {},
+    overflow: {},
+    discharged: {},
+    intensive: {},
+    dead: {},
+  }
+
+  const push = <Sub extends NumberRecordKeys<SimulationTimePoint>>(sub: Sub, age: string, delta: number) => {
+    // To get TS to type check this function, we need first assign newPop[sub] to Record<string, number>
+    // There is possibly a better solution
+    const record: Record<string, number> = newPop[sub]
+    record[age] = pop[sub][age] + delta
+  }
```

Comment:
This function was tricky to type check; however, other than splitting the body into two lines, nothing else changed.